tests/integration/test_user_management.py

```diff
@@ -0,0 +1,163 @@
+
+from . import BaseIntegrationTest, default_users
+from datameta import models
+from datameta.api import base_url
+
+from .utils import get_user, get_group
+class TestUserManagement(BaseIntegrationTest):
+    @parameterized.expand([
+        ("self_name_change", "user_a", "", "changed_by_user_a", 204),
+        ("other_name_change_by_regular_user", "user_a", "user_c", "changed_by_user_a", 403),
+        ("other_name_change_by_own_group_admin", "group_x_admin", "user_a", "changed_by_group_x_admin", 204),
+        ("other_name_change_by_other_group_admin", "group_x_admin", "user_c", "changed_by_group_x_admin", 403),
+        ("other_name_change_by_admin", "admin", "user_a", "changed_by_admin", 204),
+    ])
+    def test_name_change(self, testname:str, executing_user:str, target_user:str, new_name:str, expected_response:int):
+        user = self.users[executing_user]
+        target_user = self.users[target_user] if target_user else user
+
+        request_body = {
+            "name": new_name
+        }
+
+        req_json = {
+            "headers": user.auth.header,
+            "params": request_body,
+            "status": expected_response
+        }
+
+        response = self.testapp.put_json(
+            f"{base_url}/users/{target_user.uuid}",
+            **req_json
+        )
+
+        db_user = get_user(self.session_factory, target_user)
+        expected_outcome = any([
+            expected_response == 204 and db_user["fullname"] == new_name,
+            expected_response == 403 and db_user["fullname"] == target_user.fullname
+        ])
+        
+        assert expected_outcome, f"{testname} failed"
+
+    @parameterized.expand([
+        ("self_group_change", "user_a", "", "group_y", 403),
+        ("self_group_change_by_group_admin", "group_x_admin", "", "group_y", 403),
+        ("other_group_change_by_group_admin", "group_x_admin", "user_a", "group_y", 403),
+        ("other_group_change_by_other_group_admin", "group_x_admin", "user_c", "group_x", 403),
+        ("self_group_change_by_admin", "admin", "", "group_y", 204),
+    ])
+    def test_group_change(self, testname:str, executing_user:str, target_user:str, new_group:str, expected_response:int):
+        def process_request(user, target_user, new_group_uuid, expected_response):
+            req_json = {
+                "headers": user.auth.header,
+                "params": {"groupId": new_group_uuid},
+                "status": expected_response
+            }
+
+            response = self.testapp.put_json(
+                f"{base_url}/users/{target_user.uuid}",
+                **req_json
+            )
+
+            db_user = get_user(self.session_factory, target_user)
+            expected_outcome = any([
+                expected_response == 204 and db_user["group_uuid"] == new_group_uuid,
+                expected_response == 403 and db_user["group_uuid"] != new_group_uuid
+            ])
+        
+            assert expected_outcome, f"{testname} failed."
```

Comment:
You can move this out of the function and define as regular class method. It will not be recognized as test if it does start with "test_". Same for the below functions.

Reply:
I've done this before, but abandoned this. Nevertheless, I have now put some more refactoring work into it.